### ISSUE TITLE
Feature request: Adding support for scala Long's

### ISSUE THREAD

**Issue by saraswat** [epfl-lara/stainless#776]
For industrial projects it is useful to work with Longs as well as Ints (and BigInts).

Seems like it should be easy to add in support? 

If someone from the team can outline the process for Stainless blessing existing core classes (what needs to change, where), we could try to follow that and submit a pull request..I am poking around, but am sure we will be able to work faster with some guidance.

**Comment by samarion** [comment]
I believe in this instance that it's rather the documentation which isn't up to date wrt the implementation. [CodeExtraction](https://github.com/epfl-lara/stainless/blob/dae38d196c2a9f097391886fc12404bccf5b7d5b/frontends/scalac/src/main/scala/stainless/frontends/scalac/CodeExtraction.scala#L1629) seems to be correctly extracting the `Byte`, `Long`, etc. types. Do you have a snippet where extraction/verification is failing?

**Comment by saraswat** [comment]
It does fail for `Float`. I have tried `Byte` and `Long`. Not yet tried `Short`. 

Here is `FloatTest` -- tracebacks because of missing dependencies on `Float$0`.

```
import stainless.lang._
import stainless.collection._
import stainless.annotation._

import stainless.collection.List
// This does not work in stainless

case class CheckFloat(x: Float, a:Float) {
/*
  def i: Int = x.toInt

  // def bitops: List[Int] = List(x& a, x% a)


  def arith: List[Float] =
    List(x*a,
      x-a,
      x+a,
      // x / a,
      -x)

  def div: Float = {
    require(a != 0)
    x * a
  }

 def comp: List[Boolean] =
    List(x == a, x != a,  x < a, x <= a, x > a, x >= a)
 */
}
```

**Comment by saraswat** [comment]
`Short` works.

Where works means: no traceback, VCs are generated and checked. (May fail for divide by  zero.)

```
import stainless.lang._
import stainless.collection._
import stainless.annotation._

import stainless.collection.List
// This does not work in stainless

case class CheckFloat(x: Short, a:Short) {

    def i: Int = x.toInt

    def bitops: List[Int] = List(x& a, x% a)


    def arith: List[Int] =
      List(x*a,
        x-a,
        x+a,
        // x / a,
        -x)

    def div: Int = {
      require(a != 0)
      x / a
    }

   def comp: List[Boolean] =
      List(x == a, x != a,  x < a, x <= a, x > a, x >= a)

}

```

**Comment by samarion** [comment]
There's unfortunately no support for floating-point arithmetic in Stainless. I don't think there's any technical reason to not have it, but rather it wasn't clear what (or how) you would want to provide as specifications for floating-point operations where error bounds and stability become more relevant.

In order to add support for `Float`, the main change would need to take place inside Inox. We would need to provide a `FPType` similarly to the [BVType](https://github.com/epfl-lara/inox/blob/dc719cc9602df4b47fe20150a133910bc2b34dd6/src/main/scala/inox/ast/Types.scala#L67) in Inox and add support for encoding floating-point operations in the various solver backends. Then, we would just need to extend Stainless code extraction to extract floating-point types similarly to how bitvector types (`Long`, `Short`, etc.) are extracted.

**Comment by saraswat** [comment]
Indeed, correctness for `Float` is an issue in its own right. However, when getting existing code to work in `Stainless` it is very valuable to even have a minimal level of support for classes such as `Float`-- no extra VCs are generated, just code with `Float` compiles in `Stainless` if it compiles in `scalac` (type-safety).

So I guess I see three levels of support:

- "Normal" `Stainless` support. Source code needs to be available. `Stainless` will check that methods terminate, check `require`, `assert` and `ensuring` clauses, check exhaustiveness of `match` statements, divide by zero and other runtime exceptions. 
- "Pass through" (not a great phrase) support. Code may be available as byte code or source. `Stainless` only offers the guarantees of `scalac`. In particular, whether or not operations terminate is left up to the user to validate. Perhaps to ensure the user understands that they may introduce unsoundness in this way, have them mark the type in some way, e.g. `@passthrough`. Exactly where too do this remains a question -- `@extern` does not work well because it is a point-of-use annotation, whereas we want a point-of-definition hook.  
- "Opaque" or "extern" support. Code may be available as byte code or source. `Stainless` considers these as opaque -- `Stainless` could cannot invoke any type-specific operations on an instance of this type.

`Float` should be in the "pass through" category. (It is possible to think of better static checks, but they are not yet developed; over time `Float` could move to "normal" support.)
